File: Sources/BraintreePaymentFlow/BTLocalPaymentRequest.m
#import "BTPaymentFlowClient_Internal.h"
#import "BTPaymentFlowClient+LocalPayment_Internal.h"
#import <SafariServices/SafariServices.h>

//Objective-C Module imports
#if __has_include(<Braintree/BraintreePaymentFlow.h>) // CocoaPods
#import <Braintree/BTLocalPaymentRequest.h>
#import <Braintree/BTConfiguration+LocalPayment.h>
#import <Braintree/BTLocalPaymentResult.h>
#import <Braintree/BraintreeCore-Swift.h>

#elif SWIFT_PACKAGE                                   // SPM
#import <BraintreePaymentFlow/BTLocalPaymentRequest.h>
#import <BraintreePaymentFlow/BTConfiguration+LocalPayment.h>
#import <BraintreePaymentFlow/BTLocalPaymentResult.h>
#import <BraintreeCore/BraintreeCore-Swift.h>

#else                                                 // Carthage and Local Builds
#import <BraintreePaymentFlow/BTLocalPaymentRequest.h>
#import <BraintreePaymentFlow/BTConfiguration+LocalPayment.h>
#import <BraintreePaymentFlow/BTLocalPaymentResult.h>
#import <BraintreeCore/BraintreeCore-Swift.h>
#endif


//Swift Module imports
#if __has_include(<Braintree/Braintree-Swift.h>)      // CocoaPods-generated Swift Header
#import <Braintree/Braintree-Swift.h>

#elif SWIFT_PACKAGE                                   // SPM
/* Use @import for SPM support
 * See https://forums.swift.org/t/using-a-swift-package-in-a-mixed-swift-and-objective-c-project/27348
 */
@import BraintreeDataCollector;
@import BraintreeCore;

#elif __has_include("Braintree-Swift.h")              // CocoaPods for ReactNative
/* Use quoted style when importing Swift headers for ReactNative support
 * See https://github.com/braintree/braintree_ios/issues/671
 */
#import "Braintree-Swift.h"

#else                                                 // Carthage and Local Builds
#import <BraintreeDataCollector/BraintreeDataCollector-Swift.h>
#import <BraintreeCore/BraintreeCore-Swift.h>
#endif


@interface BTLocalPaymentRequest ()

@property (nonatomic, copy, nullable) NSString *paymentID;
@property (nonatomic, weak) id<BTPaymentFlowClientDelegate> paymentFlowClientDelegate;
@property (nonatomic, strong) NSString *correlationID;

@end

@implementation BTLocalPaymentRequest

- (void)handleRequest:(BTPaymentFlowRequest *)request client:(BTAPIClient *)apiClient paymentClientDelegate:(id<BTPaymentFlowClientDelegate>)delegate {
    self.paymentFlowClientDelegate = delegate;
    BTLocalPaymentRequest *localPaymentRequest = (BTLocalPaymentRequest *)request;
    [apiClient fetchOrReturnRemoteConfiguration:^(__unused BTConfiguration *configuration, NSError *configurationError) {
        if (configurationError) {
            [delegate onPaymentComplete:nil error:configurationError];
            return;
        }
        
        BTDataCollector *dataCollector = [[BTDataCollector alloc] initWithAPIClient:apiClient];
        self.correlationID = [dataCollector clientMetadataID:nil];

        NSError *integrationError;

        if ([self.paymentFlowClientDelegate returnURLScheme] == nil || [[self.paymentFlowClientDelegate returnURLScheme] isEqualToString:@""]) {
            NSLog(@"%@ Local Payment requires a return URL scheme to be configured via [BTAppContextSwitcher setReturnURLScheme:]", [BTLogLevelDescription stringFor:BTLogLevelCritical]);
            integrationError = [NSError errorWithDomain:BTPaymentFlowErrorDomain
                                                 code:BTPaymentFlowErrorTypeInvalidReturnURL
                                             userInfo:@{NSLocalizedDescriptionKey: @"UIApplication failed to perform app or browser switch."}];
        } else if (![configuration isLocalPaymentEnabled]) {
            NSLog(@"%@ Enable PayPal for this merchant in the Braintree Control Panel to use Local Payments.", [BTLogLevelDescription stringFor:BTLogLevelCritical]);
            integrationError = [NSError errorWithDomain:BTPaymentFlowErrorDomain
                                                 code:BTPaymentFlowErrorTypeDisabled
                                             userInfo:@{NSLocalizedDescriptionKey: @"Enable PayPal for this merchant in the Braintree Control Panel to use Local Payments."}];
        } else if (localPaymentRequest.localPaymentFlowDelegate == nil) {
            NSLog(@"%@ BTLocalPaymentRequest localPaymentFlowDelegate can not be nil.", [BTLogLevelDescription stringFor:BTLogLevelCritical]);
            integrationError = [NSError errorWithDomain:BTPaymentFlowErrorDomain
                                                 code:BTPaymentFlowErrorTypeIntegration
                                             userInfo:@{NSLocalizedDescriptionKey: @"Failed to begin payment flow: BTLocalPaymentRequest localPaymentFlowDelegate can not be nil."}];
        } else if (localPaymentRequest.amount == nil || localPaymentRequest.paymentType == nil) {
            NSLog(@"%@ BTLocalPaymentRequest amount and paymentType can not be nil.", [BTLogLevelDescription stringFor:BTLogLevelCritical]);
            integrationError = [NSError errorWithDomain:BTPaymentFlowErrorDomain
                                                 code:BTPaymentFlowErrorTypeIntegration
                                             userInfo:@{NSLocalizedDescriptionKey: @"Failed to begin payment flow: BTLocalPaymentRequest amount and paymentType can not be nil."}];
        }
        
        if (integrationError != nil) {
            [delegate onPaymentComplete:nil error:integrationError];
            return;
        }

        NSMutableDictionary *params = [@{
                                 @"amount": localPaymentRequest.amount,
                                 @"funding_source": localPaymentRequest.paymentType,
                                 @"intent": @"sale"
                                 } mutableCopy];

        params[@"return_url"] = [NSString stringWithFormat:@"%@%@", [delegate returnURLScheme], @"://x-callback-url/braintree/local-payment/success"];
        params[@"cancel_url"] = [NSString stringWithFormat:@"%@%@", [delegate returnURLScheme], @"://x-callback-url/braintree/local-payment/cancel"];

        if (localPaymentRequest.paymentTypeCountryCode) {
            params[@"payment_type_country_code"] = localPaymentRequest.paymentTypeCountryCode;
        }

        if (localPaymentRequest.address) {
            params[@"line1"] = localPaymentRequest.address.streetAddress;
            params[@"line2"] = localPaymentRequest.address.extendedAddress;
            params[@"city"] = localPaymentRequest.address.locality;
            params[@"state"] = localPaymentRequest.address.region;
            params[@"postal_code"] = localPaymentRequest.address.postalCode;
            params[@"country_code"] = localPaymentRequest.address.countryCodeAlpha2;
        }

        if (localPaymentRequest.currencyCode) {
            params[@"currency_iso_code"] = localPaymentRequest.currencyCode;
        }

        if (localPaymentRequest.givenName) {
            params[@"first_name"] = localPaymentRequest.givenName;
        }

        if (localPaymentRequest.surname) {
            params[@"last_name"] = localPaymentRequest.surname;
        }

        if (localPaymentRequest.email) {
            params[@"payer_email"] = localPaymentRequest.email;
        }

        if (localPaymentRequest.phone) {
            params[@"phone"] = localPaymentRequest.phone;
        }

        if (localPaymentRequest.merchantAccountID) {
            params[@"merchant_account_id"] = localPaymentRequest.merchantAccountID;
        }

        if (localPaymentRequest.bic) {
            params[@"bic"] = localPaymentRequest.bic;
        }

        NSMutableDictionary *experienceProfile = [@{
            @"no_shipping": @(!localPaymentRequest.isShippingAddressRequired)
        } mutableCopy];

        if (localPaymentRequest.displayName) {
            experienceProfile[@"brand_name"] = localPaymentRequest.displayName;
        }
        
        params[@"experience_profile"] = [experienceProfile copy];

        [apiClient POST:@"v1/local_payments/create"
                   parameters:params
                   completion:^(BTJSON *body, __unused NSHTTPURLResponse *response, NSError *error) {
             if (!error) {
                 self.paymentID = [body[@"paymentResource"][@"paymentToken"] asString];
                 NSString *approvalUrl = [body[@"paymentResource"][@"redirectUrl"] asString];
                 NSURL *url = [NSURL URLWithString:approvalUrl];

                 if (self.paymentID && url) {
                     [self.localPaymentFlowDelegate localPaymentStarted:self paymentID:self.paymentID start:^{
                         [delegate onPaymentWithURL:url error:error];
                     }];
                 } else {
                     NSLog(@"%@ Payment cannot be processed: the redirectUrl or paymentToken is nil.  Contact Braintree support if the error persists.", [BTLogLevelDescription stringFor:BTLogLevelCritical]);
                     NSError *error = [NSError errorWithDomain:BTPaymentFlowErrorDomain
                                                          code:BTPaymentFlowErrorTypeAppSwitchFailed
                                                      userInfo:@{NSLocalizedDescriptionKey: @"Payment cannot be processed: the redirectUrl or paymentToken is nil.  Contact Braintree support if the error persists."}];
                     [delegate onPaymentComplete:nil error:error];
                     return;
                 }
             } else {
                 if (error.code == BTCoreConstants.networkConnectionLostCode) {
                     [apiClient sendAnalyticsEvent:@"ios.local-payment-methods.network-connection.failure"];
                 }

                 [delegate onPaymentWithURL:nil error:error];
             }
         }];
    }];
}

- (void)handleOpenURL:(__unused NSURL *)url {
    if ([url.host isEqualToString:@"x-callback-url"] && [url.path hasPrefix:@"/braintree/local-payment/cancel"]) {
        // canceled
        NSError *error = [NSError errorWithDomain:BTPaymentFlowErrorDomain
                                             code:BTPaymentFlowErrorTypeCanceled
                                         userInfo:@{}];
        [self.paymentFlowClientDelegate onPaymentComplete:nil error:error];
    } else {
        // success
        NSMutableDictionary *parameters = [NSMutableDictionary dictionary];
        parameters[@"paypal_account"] = [@{} mutableCopy];
        parameters[@"paypal_account"][@"response"] = @{ @"webURL": url.absoluteString };
        parameters[@"paypal_account"][@"response_type"] = @"web";
        parameters[@"paypal_account"][@"options"] = @{ @"validate": @NO };
        parameters[@"paypal_account"][@"intent"] = @"sale";

        if (self.correlationID) {
            parameters[@"paypal_account"][@"correlation_id"] = self.correlationID;
        }

        if (self.merchantAccountID) {
            parameters[@"merchant_account_id"] = self.merchantAccountID;
        }

        BTClientMetadata *metadata =  self.paymentFlowClientDelegate.apiClient.metadata;
        parameters[@"_meta"] = @{
                                 @"source" : metadata.sourceString,
                                 @"integration" : metadata.integrationString,
                                 @"sessionId" : metadata.sessionID,
                                 };

        [self.paymentFlowClientDelegate.apiClient POST:@"/v1/payment_methods/paypal_accounts"
                  parameters:parameters
                  completion:^(BTJSON *body, __unused NSHTTPURLResponse *response, NSError *error)
         {
             if (error) {
                 if (error.code == BTCoreConstants.networkConnectionLostCode) {
                     [self.paymentFlowClientDelegate.apiClient sendAnalyticsEvent:@"ios.local-payment-methods.network-connection.failure"];
                 }
                 [self.paymentFlowClientDelegate onPaymentComplete:nil error:error];
                 return;
             } else {
                 BTJSON *payPalAccount = body[@"paypalAccounts"][0];
                 NSString *nonce = [payPalAccount[@"nonce"] asString];
                 NSString *type = [payPalAccount[@"type"] asString];

                 BTJSON *details = payPalAccount[@"details"];

                 NSString *email = [details[@"email"] asString];
                 NSString *clientMetadataID = [details[@"correlationId"] asString];
                 // Allow email to be under payerInfo
                 if ([details[@"payerInfo"][@"email"] isString]) {
                     email = [details[@"payerInfo"][@"email"] asString];
                 }

                 NSString *firstName = [details[@"payerInfo"][@"firstName"] asString];
                 NSString *lastName = [details[@"payerInfo"][@"lastName"] asString];
                 NSString *phone = [details[@"payerInfo"][@"phone"] asString];
                 NSString *payerID = [details[@"payerInfo"][@"payerId"] asString];

                 BTPostalAddress *shippingAddress = [self.class shippingOrBillingAddressFromJSON:details[@"payerInfo"][@"shippingAddress"]];
                 BTPostalAddress *billingAddress = [self.class shippingOrBillingAddressFromJSON:details[@"payerInfo"][@"billingAddress"]];
                 if (!shippingAddress) {
                     shippingAddress = [self.class accountAddressFromJSON:details[@"payerInfo"][@"accountAddress"]];
                 }

                 BTLocalPaymentResult *tokenizedLocalPayment = [[BTLocalPaymentResult alloc] initWithNonce:nonce
                                                                                                      type:type
                                                                                                      email:email
                                                                                                  firstName:firstName
                                                                                                   lastName:lastName
                                                                                                      phone:phone
                                                                                             billingAddress:billingAddress
                                                                                            shippingAddress:shippingAddress
                                                                                           clientMetadataID:clientMetadataID
                                                                                                    payerID:payerID];
                 [self.paymentFlowClientDelegate onPaymentComplete:tokenizedLocalPayment error:nil];
             }
         }];
    }
}

+ (BTPostalAddress *)accountAddressFromJSON:(BTJSON *)addressJSON {
    if (!addressJSON.isObject) {
        return nil;
    }

    BTPostalAddress *address = [[BTPostalAddress alloc] init];
    address.recipientName = [addressJSON[@"recipientName"] asString]; // Likely to be nil
    address.streetAddress = [addressJSON[@"street1"] asString];
    address.extendedAddress = [addressJSON[@"street2"] asString];
    address.locality = [addressJSON[@"city"] asString];
    address.region = [addressJSON[@"state"] asString];
    address.postalCode = [addressJSON[@"postalCode"] asString];
    address.countryCodeAlpha2 = [addressJSON[@"country"] asString];

    return address;
}

+ (BTPostalAddress *)shippingOrBillingAddressFromJSON:(BTJSON *)addressJSON {
    if (!addressJSON.isObject) {
        return nil;
    }

    BTPostalAddress *address = [[BTPostalAddress alloc] init];
    address.recipientName = [addressJSON[@"recipientName"] asString]; // Likely to be nil
    address.streetAddress = [addressJSON[@"line1"] asString];
    address.extendedAddress = [addressJSON[@"line2"] asString];
    address.locality = [addressJSON[@"city"] asString];
    address.region = [addressJSON[@"state"] asString];
    address.postalCode = [addressJSON[@"postalCode"] asString];
    address.countryCodeAlpha2 = [addressJSON[@"countryCode"] asString];

    return address;
}

- (BOOL)canHandleAppSwitchReturnURL:(NSURL *)url {
    return [url.host isEqualToString:@"x-callback-url"] && [url.path hasPrefix:@"/braintree/local-payment"];
}

- (NSString *)paymentFlowName {
    NSString *paymentType = self.paymentType != nil ? [self.paymentType lowercaseString] : @"unknown";
    return [NSString stringWithFormat:@"%@.local-payment", paymentType];
}

@end

File: Sources/BraintreePaymentFlow/BTLocalPaymentRequest.swift
import Foundation

#if canImport(BraintreeCore)
import BraintreeCore
#endif

#if canImport(BraintreeDataCollector)
import BraintreeDataCollector
#endif

/// Used to initialize a local payment flow
@objcMembers public class BTLocalPaymentRequest: NSObject {
    
    // MARK: - Public Properties
    
    /// The type of payment.
    public var paymentType: String?
    
    ///  The country code of the local payment.
    ///
    ///  This value must be one of the supported country codes for a given local payment type listed at the link below. For local payments supported in multiple countries, this value may determine which banks are presented to the customer.
    ///
    /// https://developer.paypal.com/braintree/docs/guides/local-payment-methods/client-side-custom/ios/v5#invoke-payment-flow
    public var paymentTypeCountryCode: String?
    
    /// Optional: The address of the customer. An error will occur if this address is not valid.
    public var merchantAccountID: String?
    
    /// Optional: The address of the customer. An error will occur if this address is not valid.
    public var address: BTPostalAddress?
    
    /// The amount for the transaction.
    public var amount: String?
    
    /// Optional: A valid ISO currency code to use for the transaction. Defaults to merchant currency code if not set.
    public var currencyCode: String?
    
    /// Optional: The merchant name displayed inside of the local payment flow.
    public var displayName: String?
    
    /// Optional: Payer email of the customer.
    public var email: String?
    
    /// Optional: Given (first) name of the customer.
    public var givenName: String?
    
    /// Optional: Surname (last name) of the customer.
    public var surname: String?
    
    /// Optional: Phone number of the customer.
    public var phone: String?
    
    ///  Indicates whether or not the payment needs to be shipped. For digital goods, this should be false. Defaults to false.
    public var isShippingAddressRequired: Bool = false
    
    /// Optional: Bank Identification Code of the customer (specific to iDEAL transactions).
    public var bic: String?
    
    public weak var localPaymentFlowDelegate: BTLocalPaymentRequestDelegate?
    
    // MARK: - Internal Properties
    
    var paymentID: String?
    weak var paymentFlowClientDelegate: BTPaymentFlowClientDelegate?
    var correlationID: String?
}

// MARK: - BTPaymentFlowRequestDelegate Protocol Conformance

extension BTLocalPaymentRequest: BTPaymentFlowRequestDelegate {

    /// :nodoc:
    public func handle(
        _ request: BTLocalPaymentRequest,
        client apiClient: BTAPIClient,
        paymentClientDelegate delegate: BTPaymentFlowClientDelegate
    ) {
        paymentFlowClientDelegate = delegate

        apiClient.fetchOrReturnRemoteConfiguration { configuration, error in
            if let error {
                delegate.onPaymentComplete(nil, error: error)
            }

            let dataCollector = BTDataCollector(apiClient: apiClient)
            self.correlationID = dataCollector.clientMetadataID(nil)

            guard let configuration else {
                delegate.onPaymentComplete(nil, error: BTLocalPaymentError.fetchConfigurationFailed)
                return
            }
            
            if !configuration.isLocalPaymentEnabled {
                NSLog("%@ Enable PayPal for this merchant in the Braintree Control Panel to use Local Payments.", BTLogLevelDescription.string(for: .critical))
                delegate.onPaymentComplete(nil, error: BTLocalPaymentError.disabled)
                return
            } else if (request.localPaymentFlowDelegate == nil) {
                NSLog("%@ BTLocalPaymentRequest localPaymentFlowDelegate can not be nil.", BTLogLevelDescription.string(for: .critical))
                delegate.onPaymentComplete(nil, error: BTLocalPaymentError.integration)
                return
            } else if (request.amount == nil || (request.paymentType == nil)) {
                NSLog("%@ BTLocalPaymentRequest amount and paymentType can not be nil.", BTLogLevelDescription.string(for: .critical))
                delegate.onPaymentComplete(nil, error: BTLocalPaymentError.integration)
                return
            }

            var params: [String: Any] = [
                "amount": request.amount ?? "",
                "funding_source": request.paymentType ?? "",
                "intent": "sale",
                "return_url": "\(BTCoreConstants.callbackURLScheme)://x-callback-url/braintree/local-payment/success",
                "cancel_url": "\(BTCoreConstants.callbackURLScheme)://x-callback-url/braintree/local-payment/cancel"
            ]

            if let countryCode = request.paymentTypeCountryCode {
                params["payment_type_country_code"] = countryCode
            }

            if let address = request.address {
                params["line1"] = address.streetAddress
                params["line2"] = address.extendedAddress
                params["city"] = address.locality
                params["state"] = address.region
                params["postal_code"] = address.postalCode
                params["country_code"] = address.countryCodeAlpha2
            }

            if let currencyCode = request.currencyCode {
                params["currency_iso_code"] = currencyCode
            }

            if let givenName = request.givenName {
                params["first_name"] = givenName
            }

            if let surname = request.surname {
                params["last_name"] = surname
            }

            if let email = request.email {
                params["payer_email"] = email
            }

            if let phone = request.phone {
                params["phone"] = phone
            }

            if let merchantAccountID = request.merchantAccountID {
                params["merchant_account_id"] = merchantAccountID
            }

            if let bic = request.bic {
                params["bic"] = bic
            }

            var experienceProfile: [String: Any] = [
                "no_shipping": !request.isShippingAddressRequired,
            ]

            if let displayName = request.displayName {
                experienceProfile["brand_name"] = displayName
            }

            params["experience_profile"] = experienceProfile

            apiClient.post("v1/local_payments/create", parameters: params) { body, response, error in
                if let error {
                    if (error as NSError).code == BTCoreConstants.networkConnectionLostCode {
                        apiClient.sendAnalyticsEvent("ios.local-payment-methods.network-connection.failure")
                    }
                    
                    delegate.onPayment(with: nil, error: error)
                    return
                }
                
                if let paymentID = body?["paymentResource"]["paymentToken"].asString(),
                   let approvalURLString = body?["paymentResource"]["redirectUrl"].asString(),
                   let url = URL(string: approvalURLString) {
                    
                    self.localPaymentFlowDelegate?.localPaymentStarted(self, paymentID: paymentID, start: {
                        delegate.onPayment(with: url, error: error)
                    })
                } else {
                    NSLog("%@ Payment cannot be processed: the redirectUrl or paymentToken is nil.  Contact Braintree support if the error persists.", BTLogLevelDescription.string(for: .critical))
                    delegate.onPaymentComplete(nil, error: BTLocalPaymentError.appSwitchFailed)
                    return
                }
            }
        }
    }

    /// :nodoc:
    public func handleOpen(_ url: URL) {
        if url.host == "x-callback-url" && url.path.hasPrefix("/braintree/local-payment/cancel") {
            // canceled case
            paymentFlowClientDelegate?.onPaymentComplete(nil, error: BTLocalPaymentError.canceled(paymentFlowName()))
            
        } else {
            // success case
            var params: [String: Any] = [:]
            
            var paypalAccount: [String: Any] = [:]
            paypalAccount["response"] = ["webURL": url.absoluteString]
            paypalAccount["response_type"] = "web"
            paypalAccount["options"] = ["validate": false]
            paypalAccount["intent"] = "sale"

            if let correlationID {
                paypalAccount["correlation_id"] = correlationID
            }

            if let merchantAccountID {
                params["merchant_account_id"] = merchantAccountID
            }
            
            params["paypal_account"] = paypalAccount
            
            let metadata = paymentFlowClientDelegate?.apiClient().metadata
            params["_meta"] = [
                "source": metadata?.sourceString,
                "integraton": metadata?.integrationString,
                "sessionId": metadata?.sessionID
            ]
            
            paymentFlowClientDelegate?.apiClient().post("/v1/payment_methods/paypal_accounts", parameters: params, completion: { [weak self] body, response, error in
                guard let self else { return }
                
                if let error {
                    if (error as NSError).code == BTCoreConstants.networkConnectionLostCode {
                        self.paymentFlowClientDelegate?.apiClient().sendAnalyticsEvent("ios.local-payment-methods.network-connection.failure")
                    }
                    
                    self.paymentFlowClientDelegate?.onPaymentComplete(nil, error: error)
                    return
                } else {
                    guard let body else {
                        self.paymentFlowClientDelegate?.onPaymentComplete(nil, error: BTLocalPaymentError.noAccountData)
                        return
                    }
                    
                    guard let tokenizedLocalPayment = BTLocalPaymentResult(json: body) else {
                        self.paymentFlowClientDelegate?.onPaymentComplete(nil, error: BTLocalPaymentError.failedToCreateNonce)
                        return
                    }
                    
                    self.paymentFlowClientDelegate?.onPaymentComplete(tokenizedLocalPayment, error: nil)
                }
            })
        }
    }

    /// :nodoc:
    public func paymentFlowName() -> String {
        let paymentType = paymentType?.lowercased() ??  "unknown"
        return "\(paymentType).local-payment"
    }
}
